src/routes/safe/components/Apps/AddAppForm.tsx

```diff
@@ -0,0 +1,203 @@
+    }
+  })
+  return exists ? 'This app is already registered.' : undefined
+})
+
+const getIpfsLinkFromEns = memoize(async (name) => {
+  try {
+    const content = await getContentFromENS(name)
+    if (content && content.protocolType === 'ipfs') {
+      return `${process.env.REACT_APP_IPFS_GATEWAY}/${content.decoded}/`
+    }
+  } catch (error) {
+    console.error(error)
+    return undefined
+  }
+})
+
+const getUrlFromFormValue = memoize(async (value: string) => {
+  const isUrlValid = isURLValid(value)
+  let ensContent
+  if (!isUrlValid) {
+    ensContent = await getIpfsLinkFromEns(value)
+  }
+
+  if (!isUrlValid && ensContent === undefined) {
+    return undefined
+  }
+  return isUrlValid ? value : ensContent
+})
+
+const curriedSafeAppValidator = memoize((appList) => async (value: string) => {
+  const url = await getUrlFromFormValue(value)
+
+  if (!url) {
+    return 'Provide a valid url or ENS name.'
+  }
+
+  const appExistsRes = appExistsValidator(appList, url)
+  if (appExistsRes) {
+    return appExistsRes
+  }
+
+  const appInfo = await getAppInfoFromUrl(url)
+  if (appInfo.error) {
+    return 'This is not a valid Safe app.'
+  }
+})
+
+const composeValidatorsApps = (...validators: Function[]) => (value, values, meta) => {
+  if (!meta.modified) {
+    return
+  }
+  return validators.reduce((error, validator) => error || validator(value), undefined)
+}
+
+type Props = {
+  formId: string
+  appList: Array<SafeApp>
+  closeModal: () => void
+  onAppAdded: (app: SafeApp) => void
+  setIsSubmitDisabled: (status: boolean) => void
+}
+
+const AddAppForm = ({ appList, formId, closeModal, onAppAdded, setIsSubmitDisabled }: Props) => {
+  const [appInfo, setAppInfo] = useState<SafeApp>(APP_INFO)
+  const safeAppValidator = curriedSafeAppValidator(appList)
+
+  const onFormStatusChange = async ({ pristine, valid, validating, values, errors }) => {
+    if (!pristine) {
+      setIsSubmitDisabled(validating || !valid)
+    }
+
+    if (validating) {
+      return
+    }
+
+    if (!values.appUrl || !values.appUrl.length || errors.appUrl) {
+      setAppInfo(APP_INFO)
+      return
+    }
+
+    const url = await getUrlFromFormValue(values.appUrl)
+    const appInfo = await getAppInfoFromUrl(url)
+    setAppInfo({ ...appInfo })
+  }
+
+  const handleSubmit = () => {
+    closeModal()
+    onAppAdded(appInfo)
+  }
+
+  return (
+    <GnoForm
+      initialValues={{
+        appUrl: '',
+        agreed: false,
+      }}
+      // submit is triggered from ManageApps Component
+      onSubmit={handleSubmit}
+      testId={formId}
+    >
+      {() => (
+        <>
+          <StyledText size="xl">Add custom app</StyledText>
+          <DebounceValidationField
+            component={TextField}
+            label="App URL"
+            name="appUrl"
+            placeholder="App URL"
+            type="text"
+            validate={composeValidatorsApps(required, safeAppValidator)}
```

Comment:
I think I miss the logic behind using compose there, so in safeAppValidator you take all the validators and put them into one.
Why couldn't it be
```
composeValidatorApps(required, validUrlOrENSname, uniqueAppValidator, validAppValidator)
```
or the other way around - moving required validator to safeAppValidator?

I would prefer the first way to keep the consistency with how we do it in other parts of the app.

Reply:
react-final-form just allow to have 1 validator returning a promise and it must be at the end. That's why I had to mix the validators.

Reply:
Where did you find that?

I could successfully use 2 validators using an example for hybrid sync/async validation example from their doc

doc -> https://final-form.org/docs/react-final-form/examples
my edited example -> https://codesandbox.io/s/react-final-form-hybrid-synchronousasynchronous-record-level-validation-example-dwxf8

Reply:
you are not using Fied validation in that example. 
https://stackoverflow.com/questions/53669263/final-form-array-fields-async-validation/53707358#53707358

Reply:
Good to know, thanks :)